File: tests/java/com/secpilot/DosageTest.java
package com.secpilot;

import com.secpilot.healthcare.Dosage;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Date;

/**
 * Tests unitaires pour le module de dosage médical
 */
class DosageTest {

    @Test
    @DisplayName("Calcule un dosage normal dans les limites")
    void testNormalDosage() {
        // 50kg à 20mg/kg = 1000mg
        double result = Dosage.calculateDosage(50, 20, "paracetamol");
        assertEquals(1000, result, 0.01);
    }

    // TEST CONTEXTUEL : Règle métier - limite de dose maximale
    @Test
    @DisplayName("RÈGLE MÉTIER : Le dosage ne peut pas dépasser la limite maximale sûre")
    void testDosageCannotExceedMaximum() {
        /**
         * Contexte : Système de santé - la sécurité patient est critique
         * Les dosages ne doivent jamais dépasser les limites maximales sûres
         */
        // 150kg à 30mg/kg = 4500mg, mais max est 4000mg
        double result = Dosage.calculateDosage(150, 30, "paracetamol");
        assertTrue(result <= 4000,
            "Le dosage " + result + "mg dépasse le maximum sûr de 4000mg");
    }

    @Test
    @DisplayName("Le dosage devrait être plafonné au maximum, pas rejeté")
    void testDosageCappedAtMaximum() {
        // 200kg à 50mg/kg = 10000mg, devrait plafonner à 4000mg
        double result = Dosage.calculateDosage(200, 50, "paracetamol");
        assertEquals(4000, result, 0.01,
            "Le dosage devrait être plafonné à la limite maximale sûre");
    }

    // TEST BUG CLASSIQUE : Division par zéro
    @Test
    @DisplayName("BUG CLASSIQUE : Une fréquence nulle devrait lever une exception")
    void testZeroFrequencyHandling() {
        /**
         * Bug : Division par zéro quand frequency est 0
         */
        assertThrows(ArithmeticException.class, () -> {
            Dosage.splitDailyDose(1000, 0);
        });
    }

    // TEST BUG CLASSIQUE : Sensibilité à la casse
    @Test
    @DisplayName("BUG CLASSIQUE : Vérification d'interaction insensible à la casse")
    void testCaseInsensitiveDrugCheck() {
        /**
         * Bug : Comparaison sensible à la casse manque les interactions
         */
        assertTrue(Dosage.checkDrugInteraction("WARFARIN", "ASPIRIN"),
            "Devrait détecter l'interaction quelle que soit la casse");
        assertTrue(Dosage.checkDrugInteraction("Warfarin", "Aspirin"),
            "Devrait détecter l'interaction avec casse mixte");
    }

    // TEST BUG CLASSIQUE : NumberFormatException
    @Test
    @DisplayName("BUG CLASSIQUE : Gestion de chaîne de dosage invalide")
    void testInvalidDosageString() {
        /**
         * Bug : Pas de try-catch pour format de nombre invalide
         */
        assertThrows(NumberFormatException.class, () -> {
            Dosage.parseDosage("invalid");
        });
        // Ou devrait retourner une valeur par défaut sensée / lever IllegalArgumentException
    }

    // TEST BUG CLASSIQUE : Mutation de Date
    @Test
    @DisplayName("BUG CLASSIQUE : Mutation de Date dans calculateNextDoseTime")
    void testDateMutation() {
        /**
         * Bug : La méthode mute l'objet Date original
         */
        Date original = new Date();
        long originalTime = original.getTime();

        Dosage.calculateNextDoseTime(original, 4);

        assertEquals(originalTime, original.getTime(),
            "La date originale ne devrait pas être mutée");
    }

    // TEST BUG CLASSIQUE : Dépassement d'entier
    @Test
    @DisplayName("BUG CLASSIQUE : Dépassement d'entier dans la validation")
    void testIntegerOverflow() {
        /**
         * Bug : dose * multiplier pourrait dépasser Integer.MAX_VALUE
         */
        boolean result = Dosage.validateDosageRange(
            Integer.MAX_VALUE / 2, 3, Integer.MAX_VALUE);

        // Si dépassement, le résultat serait incorrectement true
        assertFalse(result, "Devrait détecter que la dose dépasse le maximum même avec dépassement");
    }
}

File: tests/javascript/dosage.test.js
/**
 * Tests unitaires pour le module de dosage médical
 */
const {
    MAX_DOSES,
    calculateDosage,
    parseDosage,
    isDoseSafe,
    calculateNextDoseTime,
    validatePatientAge
} = require('../../src/javascript/healthcare/dosage');

describe('calculateDosage', () => {
    test('calcule un dosage normal dans les limites', () => {
        // 50kg à 20mg/kg = 1000mg (dans la limite de 4000mg)
        const result = calculateDosage(50, 20, 'paracetamol');
        expect(result).toBe(1000);
    });

    // TEST CONTEXTUEL : Règle métier - limite de dose maximale
    test('RÈGLE MÉTIER : le dosage ne peut pas dépasser la limite maximale sûre', () => {
        /**
         * Contexte : Système de santé - la sécurité patient est critique
         * Les dosages ne doivent jamais dépasser les limites maximales sûres
         */
        // 150kg à 30mg/kg = 4500mg, mais max est 4000mg
        const result = calculateDosage(150, 30, 'paracetamol');
        expect(result).toBeLessThanOrEqual(MAX_DOSES.paracetamol);
    });

    test('plafonne les calculs extrêmement élevés au maximum', () => {
        // 200kg à 50mg/kg = 10000mg, devrait plafonner à 4000mg
        const result = calculateDosage(200, 50, 'paracetamol');
        expect(result).toBe(MAX_DOSES.paracetamol);
    });
});

describe('parseDosage', () => {
    test('parse une chaîne numérique simple', () => {
        expect(parseDosage('500')).toBe(500);
    });

    // TEST BUG CLASSIQUE : parseInt sans radix
    test('BUG CLASSIQUE : parseInt devrait utiliser radix 10', () => {
        /**
         * Bug : parseInt('08') sans radix pourrait retourner 0 en ancien JS
         */
        expect(parseDosage('08')).toBe(8);
        expect(parseDosage('09')).toBe(9);
    });
});

describe('isDoseSafe', () => {
    test('identifie une dose dans la plage', () => {
        expect(isDoseSafe(500, 100, 1000)).toBe(true);
    });

    // TEST BUG CLASSIQUE : Comparaison de borne
    test('BUG CLASSIQUE : une dose égale à maxDose devrait être sûre', () => {
        /**
         * Bug : Utilise < au lieu de <= pour la borne supérieure
         */
        expect(isDoseSafe(1000, 100, 1000)).toBe(true);  // Exactement au max
    });

    test('identifie une dose hors plage', () => {
        expect(isDoseSafe(50, 100, 1000)).toBe(false);   // Sous le min
        expect(isDoseSafe(1001, 100, 1000)).toBe(false); // Au-dessus du max
    });
});

describe('calculateNextDoseTime', () => {
    // TEST BUG CLASSIQUE : Mutation de Date
    test('BUG CLASSIQUE : ne devrait pas muter la Date originale', () => {
        /**
         * Bug : La fonction mute l'objet Date passé
         */
        const original = new Date('2024-01-01T08:00:00');
        const originalTime = original.getTime();

        calculateNextDoseTime(original, 4);

        // L'original devrait être inchangé
        expect(original.getTime()).toBe(originalTime);
    });

    test('calcule l\'heure de prochaine dose correctement', () => {
        const lastDose = new Date('2024-01-01T08:00:00');
        const nextDose = calculateNextDoseTime(new Date(lastDose), 4);

        expect(nextDose.getHours()).toBe(12);
    });
});

describe('validatePatientAge', () => {
    test('valide un âge adulte correctement', () => {
        expect(validatePatientAge(25, 18)).toBe(true);
    });

    // TEST BUG CLASSIQUE : Truthy/falsy avec 0
    test('BUG CLASSIQUE : un âge de 0 devrait être géré correctement', () => {
        /**
         * Bug : un âge de 0 est falsy et échouerait incorrectement la validation
         */
        // L'âge 0 est un âge valide (nouveau-né), ne devrait pas être traité comme manquant
        expect(validatePatientAge(0, 0)).toBe(true);  // Nouveau-né, âge min 0
    });

    test('rejette un âge sous le minimum', () => {
        expect(validatePatientAge(5, 18)).toBe(false);
    });
});
